samples/boards/esp32/wifi_apsta_mode/src/main.c

```diff
@@ -0,0 +1,189 @@
+			sta_info->mac[2], sta_info->mac[3], sta_info->mac[4], sta_info->mac[5]);
+		break;
+	}
+	case NET_EVENT_WIFI_AP_STA_DISCONNECTED: {
+		struct wifi_ap_sta_info *sta_info = (struct wifi_ap_sta_info *)cb->info;
+
+		LOG_INF("station: " MACSTR " leave ", sta_info->mac[0], sta_info->mac[1],
+			sta_info->mac[2], sta_info->mac[3], sta_info->mac[4], sta_info->mac[5]);
+		break;
+	}
+	default:
+		break;
+	}
+}
+
+static void enable_dhcpv4_server(void)
+{
+	static struct in_addr addr;
+	static struct in_addr netmaskAddr;
+
+	if (net_addr_pton(AF_INET, WIFI_AP_IP_ADDRESS, &addr)) {
+		LOG_ERR("Invalid address: %s", WIFI_AP_IP_ADDRESS);
+		return;
+	}
+
+	if (net_addr_pton(AF_INET, WIFI_AP_NETMASK, &netmaskAddr)) {
+		LOG_ERR("Invalid netmask: %s", WIFI_AP_NETMASK);
+		return;
+	}
+
+	net_if_ipv4_set_gw(ap_iface, &addr);
+
+	if (net_if_ipv4_addr_add(ap_iface, &addr, NET_ADDR_MANUAL, 0) == NULL) {
+		LOG_ERR("unable to set IP address for AP interface");
+	}
+
+	if (!net_if_ipv4_set_netmask_by_addr(ap_iface, &addr, &netmaskAddr)) {
+		LOG_ERR("Unable to set netmask for AP interface: %s", WIFI_AP_NETMASK);
+	}
+
+	addr.s4_addr[3] += 10; /* Starting IPv4 address for DHCPv4 address pool. */
+
+	if (net_dhcpv4_server_start(ap_iface, &addr) != 0) {
+		LOG_ERR("DHCP server is not started for desired IP");
+		return;
+	}
+
+	LOG_INF("DHCPv4 server started...\n");
+}
+
+static int enable_ap_mode(void)
+{
+	if (!ap_iface) {
+		LOG_INF("AP: is not initialized");
+		return -EIO;
+	}
+
+	LOG_INF("Turning on AP Mode");
+	ap_config.ssid = (const uint8_t *)WIFI_AP_SSID;
+	ap_config.ssid_length = strlen(WIFI_AP_SSID);
+	ap_config.psk = (const uint8_t *)WIFI_AP_PSK;
+	ap_config.psk_length = strlen(WIFI_AP_PSK);
+	ap_config.channel = WIFI_CHANNEL_ANY;
+	ap_config.band = WIFI_FREQ_BAND_2_4_GHZ;
+
+	if (strlen(WIFI_AP_PSK) == 0) {
+		ap_config.security = WIFI_SECURITY_TYPE_NONE;
+	} else {
+
+		ap_config.security = WIFI_SECURITY_TYPE_PSK;
+	}
+
+	enable_dhcpv4_server();
+
+	int ret = net_mgmt(NET_REQUEST_WIFI_AP_ENABLE, ap_iface, &ap_config,
+			   sizeof(struct wifi_connect_req_params));
+	if (ret) {
+		LOG_ERR("NET_REQUEST_WIFI_AP_ENABLE failed, err: %d", ret);
+	}
+
+	return ret;
+}
+
+static int connect_to_wifi(void)
+{
+	if (!sta_iface) {
+		LOG_INF("STA: interface no initialized");
+		return -EIO;
+	}
+
+	sta_config.ssid = (const uint8_t *)WIFI_SSID;
+	sta_config.ssid_length = strlen(WIFI_SSID);
+	sta_config.psk = (const uint8_t *)WIFI_PSK;
+	sta_config.psk_length = strlen(WIFI_PSK);
+	sta_config.security = WIFI_SECURITY_TYPE_PSK;
+	sta_config.channel = WIFI_CHANNEL_ANY;
+	sta_config.band = WIFI_FREQ_BAND_2_4_GHZ;
+
+	LOG_INF("Connecting to SSID: %s\n", sta_config.ssid);
+
+	int ret = net_mgmt(NET_REQUEST_WIFI_CONNECT, sta_iface, &sta_config,
+			   sizeof(struct wifi_connect_req_params));
+	if (ret) {
+		LOG_ERR("Unable to Connect to (%s)", WIFI_SSID);
+	}
+
+	return ret;
+}
+
+int main(void)
+{
+	k_sleep(K_SECONDS(5));
```

Comment:
@zin-HXQ, I know you applied a great effort in this PR and I do appreciated it, but I feel like we need to discuss about this 5 seconds delay. If you remove it (I haven't tested the PR yet), will it fail performing next steps? I wonder whether this delay could be removed or replaced by a synchronization mechanism instead. Not opposed to keep as is and update in another PR eventually.

Reply:
No it will not fail.

If you say so we can do it in this same PR.

Reply:
If it won't fail, what is the purpose of having this delay? Just curious.

Reply:
When we started working with Wi-Fi, we were having trouble in connecting to AP. We somewhere read that you should wait for few seconds until interface is up before trying to request connect. Since then we are 5 second initial delay. Before actually running the application. 

After you asked to remove it, I tried without delay in the provided code example, it worked.

Reply:
Got it, thanks for re-testing. Sure I will try here as well soon.